### PR TITLE
Add buffer stop odd location

### LINKED ISSUE #3607: Trigger a warning if a buffer stop has a suspicious location
## Description and goal

We should have a warning if a buffer stop is located on an internal graph edge.

## Acceptance criteria

For a train : 

- [x] Remove `applicable_directions` field
- [x] Adapt infra generator
- [x] Adapt import of infra
- [x] Add the error to the infra errors
- [x] Adapt openapi
- [x] Add the check to the buffer stop

### PR DESCRIPTION
**Pull request by flomonster**
close #3607

### CODE REVIEW

File: editoast/src/generated_data/error/buffer_stops.rs [hunk 1]
```diff
@@ -60,6 +61,57 @@ fn check_out_of_range(
     }
 }
 
+/// Check the location of the buffer stop
+/// Normal location are:
+/// - On tracks that are at the end of infrastructure
+/// - The only buffer stop that protect the end of the infra
```
Comment (Tguisnet) [inline editoast/src/generated_data/error/buffer_stops.rs]: This sentence is unclear (line 67).

File: editoast/src/generated_data/error/buffer_stops.rs [hunk 2]
```diff
@@ -60,6 +61,57 @@ fn check_out_of_range(
...
+    if is_linked_end && is_linked_start {
+        return vec![InfraError::new_odd_buffer_stop_location(buffer_stop)];
+    }
+
+    let buffer_stops = infra_cache.get_track_refs_type(track_id, ObjectType::BufferStop);
+
+    // The number of buffer stops match the number of end of the track
+    if buffer_stops.len() == 1 || (buffer_stops.len() == 2 && !is_linked_end && !is_linked_start) {
+        return vec![];
+    }
```
Comment (Tguisnet) [inline editoast/src/generated_data/error/buffer_stops.rs]: Could you add a test for the case with only one track and 2 buffer stops ?

**Review by Tguisnet — CHANGES_REQUESTED** [review]

**Review by Tguisnet — DISMISSED** [review]

File: editoast/openapi.yaml [hunk 3]
```diff
@@ -1866,6 +1850,27 @@ components:
...
+        - empty_object
+        - object_out_of_path
+        - odd_buffer_stop_location
+        - missing_route
+        - unused_port
+        - duplicated_group
+        - missing_buffer_stop
+        - path_is_not_continuous
+        - overlapping_switches
+        - overlapping_track_links
```
Comment (clarani) [inline editoast/openapi.yaml]: Should we order them alphabetically ? :)

File: editoast/src/generated_data/error/buffer_stops.rs [hunk 4]
```diff
@@ -60,6 +61,58 @@ fn check_out_of_range(
...
+    let is_linked_end = graph.has_neighbour(&TrackEndpoint::new(track_id, Endpoint::End));
+
+    // The track is not at the end of the infra
+    if is_linked_end && is_linked_start {
+        return vec![InfraError::new_odd_buffer_stop_location(buffer_stop)];
+    }
+
+    let buffer_stops = infra_cache.get_track_refs_type(track_id, ObjectType::BufferStop);
+
+    // The number of buffer stops match the number of end of the track
```
Comment (clarani) [inline editoast/src/generated_data/error/buffer_stops.rs]:
```suggestion
    // The number of buffer stops matches the number of ends of the track
```

File: editoast/src/generated_data/error/buffer_stops.rs [hunk 5]
```diff
@@ -60,6 +61,58 @@ fn check_out_of_range(
...
+    if buffer_stops.len() == 1 || (buffer_stops.len() == 2 && !is_linked_end && !is_linked_start) {
+        return vec![];
+    }
+
+    // Retrieve the first and last buffer stop on the track
```
Comment (clarani) [inline editoast/src/generated_data/error/buffer_stops.rs]:
```suggestion
    // Retrieve the first and last buffer stops on the track
```

**Review by clarani — DISMISSED** [review]
Some typos and one remark on ordering, but LGTM

**Review by clarani — APPROVED** [review]
Thank you, your code is very clear and simple, very pleasant to review :D

**Pull request merged**